File: fusion/swig/tests/gqueue.lua
#!/usr/bin/env lua

-- test_gqueue.lua
local mglib = require("mglib")

print("--- Testing GQueue with Lua ---")

-- 1. Integer Queue Test
-- Input: Lua Table -> C: GIntQueue
local my_nums = {10, 20, 30, 40, 50}
print("Summing integers from Lua table: " .. table.concat(my_nums, ", "))

local total = mglib.sum_int_queue(my_nums)
print("Result from C sum_queue_int: " .. total)

-- Output: C: GIntQueue -> Lua Table
local sample_nums = mglib.get_sample_int_queue()
print("Sample integers from C: " .. table.concat(sample_nums, ", "))

-- 2. String Queue Test
-- Input: Lua Table -> C: GStringQueue
local my_strs = {"Lua", "is", "fast"}
print("\nCounting lengths of strings: " .. table.concat(my_strs, " "))

local char_count = mglib.sum_string_queue_lengths(my_strs)
print("Total characters from C: " .. char_count)

-- Output: C: GStringQueue -> Lua Table
local sample_strs = mglib.get_sample_str_queue()
print("Sample strings from C:")
for i, v in ipairs(sample_strs) do
    print("  [" .. i .. "] " .. v)
end

-- 3. Boolean Queue Test (NEW)
-- Input: Lua Table (booleans) -> C: GBoolQueue
local my_bools = {true, false, true, true, false}
print("\nTesting Boolean Queue: " .. table.concat({ "true", "false", "true", "true", "false" }, ", "))

-- Calls count_true_queue(GBoolQueue *queue) 
local true_count = mglib.count_true_queue(my_bools)
print("Number of 'true' values from C: " .. true_count)

-- Output: C: GBoolQueue -> Lua Table
local sample_bools = mglib.create_bool_queue(3) -- Creates a queue of size 3
print("Sample booleans from C:")
for i, v in ipairs(sample_bools) do
    print("  [" .. i .. "] " .. tostring(v))
end


-- 4. Float Queue Test (NEW)
-- Input: Lua Table (numbers) -> C: GFloatQueue
local my_floats = {1.5, 2.5, 3.5}
print("\nSumming floats from Lua table: 1.5, 2.5, 3.5")

-- Calls sum_float_queue(GFloatQueue *queue) 
local float_sum = mglib.sum_float_queue(my_floats)
print("Result from C sum_float_queue: " .. float_sum)

-- Output: C: GFloatQueue -> Lua Table
local sample_floats = mglib.create_float_queue(4)
print("Sample floats from C:")
for i, v in ipairs(sample_floats) do
    print("  [" .. i .. "] " .. string.format("%.2f", v))
end

File: fusion/swig/tests/gqueue.py
#!/usr/bin/env python3
import mglib

print("--- Testing GQueue with Python ---")

# 1. Integer Queue Test
# Input: Python List -> C: GIntQueue
my_nums = [10, 20, 30, 40, 50]
print(f"Summing integers from Python list: {', '.join(map(str, my_nums))}")

total = mglib.sum_int_queue(my_nums)
print(f"Result from C sum_int_queue: {total}")

# Output: C: GIntQueue -> Python List
sample_nums = mglib.get_sample_int_queue()
print(f"Sample integers from C: {', '.join(map(str, sample_nums))}")


# 2. String Queue Test
# Input: Python List -> C: GStringQueue
my_strs = ["Python", "is", "powerful"]
print(f"\nCounting lengths of strings: {' '.join(my_strs)}")

# Note: Using the function name defined in gqueue.i / gqueue.h
char_count = mglib.total_length_queue(my_strs)
print(f"Total characters from C: {char_count}")

# Output: C: GStringQueue -> Python List
sample_strs = mglib.get_sample_str_queue()
print("Sample strings from C:")
for i, v in enumerate(sample_strs):
    print(f"  [{i}] {v}")

# 3. Boolean Queue Test (NEW)
# Input: Python List (bools) -> C: GBoolQueue
my_bools = [True, False, True, True]
print(f"\nTesting Boolean Queue: {', '.join(map(str, my_bools))}")

# Calls count_true_queue(GBoolQueue *queue)
true_count = mglib.count_true_queue(my_bools)
print(f"Number of 'True' values from C: {true_count}")

# Output: C: GBoolQueue -> Python List
sample_bools = mglib.create_bool_queue(3)
print("Sample booleans from C:")
for i, v in enumerate(sample_bools):
    print(f"  [{i}] {v}")


# 4. Float Queue Test (NEW)
# Input: Python List (floats) -> C: GFloatQueue
my_floats = [1.1, 2.2, 3.3]
print(f"\nSumming floats from Python list: {', '.join(map(str, my_floats))}")

# Calls sum_float_queue(GFloatQueue *queue)
float_sum = mglib.sum_float_queue(my_floats)
print(f"Result from C sum_float_queue: {float_sum:.2f}")

# Output: C: GFloatQueue -> Python List
sample_floats = mglib.create_float_queue(4)
print("Sample floats from C:")
for i, v in enumerate(sample_floats):
    print(f"  [{i}] {v:.2f}")
